horovod/tensorflow/mpi_ops.cc

```diff
@@ -412,10 +504,13 @@ class HorovodAllreduceOp : public AsyncOpKernel {
     auto hvd_context = std::make_shared<TFOpContext>(context);
     auto hvd_tensor = std::make_shared<TFTensor>(tensor);
     auto hvd_output = std::make_shared<TFTensor>(*output);
+    TensorflowNvtxRange nvtx_range(RegisteredNvtxOp::HorovodAllreduce,
+                                   hvd_tensor->size());
     auto enqueue_result = EnqueueTensorAllreduce(
         hvd_context, hvd_tensor, hvd_output, ready_event, node_name, device,
-        [context, done](const common::Status& status) {
+        [context, done, nvtx_range](const common::Status& status) mutable {
```

Comment:
Would it be possible to move the creation of these ranges down into the Horovod backend, rather than create them within framework specific source? For example, moving the creation of these ranges into the `EnqueueTensor*` functions in `operations.cc`, and then adding the range end call to the code in `GPUOpContext::FinalizeGPUQueue` in `gpu_operations.cc`?  The benefit of this approach is that it would not require adding this NVTX code to every framework implementation individually.

Reply:
I agree that it would be nicer to not artificially couple this to the TF framework source. However, if I move the creation of these ranges somewhere else, I can no longer easily end these ranges from the callbacks because these callbacks are only defined in framework code. I would have to explicitly track the nvtx_range object somewhere (maybe in the tensor table entry?) and make sure to explicitly end it in every possible place (basically everywhere the callback is called in `operations.cc` and `gpu_operations.cc`).

How would you feel about pulling the (appropriately renamed) TensorflowNvtxRange into a common header file? I could then add those few lines to start and end ranges to `torch/mpi_ops_v2.cc` and `mxnet/mpi_ops.cc`.

Reply:
A header is one way to make it more generic, but still requires anyone who adds a new op to also make sure to add code to add this range handling to their callbacks which isn't ideal. I agree that that `nvtx_range` objects will need to be stored, and I think it is ok to add them to the tensor table entries. To deal with ending ranges where callbacks are called, maybe we can add a method to the tensor table entries like `TensorTableEntry::ExecuteCallback(Status& s)` that will end the nvtx range in the entry and then call the callback? Then, we would just replace direct calls like `e.callback(status)` to `e.ExecuteCallback(status)` to the callback with this method. 
There is one place in the code where callbacks are extracted from the entries and placed into a separate vector here: https://github.com/horovod/horovod/blob/612fe6b21aa51f7683682950a6a07d4bc6935713/horovod/common/operations.cc#L563-L569
However, the callbacks can just be called within `FinalizeTensorQueue`, since there doesn't seem to be a reason they are gathered and called outside.

Reply:
OK, you convinced me to go with your suggestion. Thanks for the pointers! In the updated version the NVTX ranges are independent of any framework-specific code. That should indeed make it easier to contribute new ops in the future.

Now I've touched more parts of the codebase than I originally set out to. Therefore I am all the more open to advice on how to name and structure things.